rfcs/0026-memory-control.md

```diff
@@ -0,0 +1,113 @@
+
+Recently, [RFC: Coarse grained resource management for batch query engine](https://github.com/risingwavelabs/rfcs/pull/11) tried to control the memory of batch executors as well, but the 1st and 2nd version of the design was controversial because some designs were too complicated or infeasible. Luckily, the RFC has been refined a lot during the discussion, which finally comes to the solution of this RFC. 
+
+## Goals
+
+- The total memory usage is controlled to prevent OOM (Out-of-Memory).
+- The streaming and batch executors should be isolated as far as possible.
+
+## Design
+
+Memory of compute node is a scarce resource. Due to the nature of streaming processing, the memory will be sooner or later consumed out. Therefore, we need to limit the memory used by each component to prevent OOM.
+
+![](images/overall-memory-allocation.drawio.svg)
+
+**This RFC will focus on the memory controlling on both streaming and batch executors**, assuming the storage, network buffer, system metadata, etc. always take a fixed size of memory.
+
+Our approach uses one background coroutine to continuously run the loop below (inspired by @yuhao-su):
+
+1. **Measure**: Get the current memory usage
+2. **Policy**: Decide whether and which component needs to release memory
+3. **Action**: Take actions to release memory
+
+### Measure: How to measure memory usage?
+
+Here we propose to use the simple and stupid way to calculate the memory usage - **estimating** the memory consumed by all the in-memory data structures and aggregating in levels. This is opposite to the solutions that rely on memory allocators to get an exact memory usage.
+
+How to estimate memory usage depends on the data structures:
+
+- For indexing structures like `HashMap`, `BTreeMap`, etc., we could estimate the memory size according to the `capacity()` or something else. 
+   - @liurenjie1024: It also seems possible to inject an `Allocator` to count the exact allocated space, as long as it exposed such generic type parameter.
+- For pure data structures like `DataChunk`, `Vec<Datum>`, etc., we could compute the exact size it used by summing up the sizes of internal arrays & variables as well as padding and other overhead.
+
+A CN-level memory manager will collect the memory usage in a fixed interval like 100ms. It collects the memory recursively regarding the hierarchy: stream/batch -> fragments -> actors -> executors. To simplify the ownership, executors' memory usage can be stored in `AtomicUsize` and be updated by executors on every `next()`.
```

Comment:
Yes I agree. Something like https://github.com/risingwavelabs/rfcs/pull/11#issuecomment-1330349179.

Reply:
I think task level's is enough?

Reply:
Yes I agree, but the problem is now the executor do not have a way to report memory usage to task level easily (e.g. How frequent executor update the memory usage number?), so it has to come from up to bottom.

Reply:
I think updating an `AtomicUsize` is enough? Passing a task level `AtomicUsize` to all executors.

Reply:
When will executor update the atomic Usize 🤔

Reply:
That's the reason why I suggest to use `Allocator`: you don't need to care about it, only attach appropriate allocator to necessary part.

Reply:
But seems like Allocator can not count all usage? 

https://github.com/risingwavelabs/risingwave/blob/ad03b2ef559f5f629313c576f64f626d264863ca/src/stream/src/executor/managed_state/join/join_entry_state.rs#L97-L102

@yuhao-su here is Allocator + estimate size.

Reply:
I guess this is some wrapper around collections such as `BTreeMap`.

Reply:
There is already existing work for `TaskLocalAlloc`: https://github.com/risingwavelabs/risingwave/pull/4785

Reply:
> When will executor update the atomic Usize 🤔

On every time it processes a DataChunk, I suppose.